File: scripts/convert_fences.py
#!/usr/bin/env python3
"""
Bulk Fence Converter for MyST Markdown

Converts MyST directives from colon fences (:::) to backtick fences (```)
to comply with MYST_CONVENTIONS.md (updated to prefer backticks for IDE support)

Usage:
    python convert_fences.py [--dry-run] [--content-dir DIR] [--file FILE]

Options:
    --dry-run       Show changes without applying them
    --content-dir   Convert all files in directory (default: content)
    --file          Convert a single file
    --verbose       Show detailed conversion info
"""

import os
import re
import sys
import argparse
from pathlib import Path
from collections import defaultdict
from typing import List, Union, DefaultDict

# Import shared utilities for logging support
from shared_utils import logger

class FenceConverter:
    """Converts colon directive fences to backtick fences."""

    # Directives that should use backtick fences
    BACKTICK_FENCE_DIRECTIVES = [
        'note', 'warning', 'important', 'tip', 'caution', 'attention',
        'danger', 'error', 'hint', 'seealso', 'admonition',
        'dropdown', 'card', 'tab-set', 'tab-item',
        'figure', 'subfigure', 'video', 'image',
        'proof', 'theorem', 'lemma', 'definition', 'example', 'exercise',
        'grid', 'column', 'margin', 'sidebar',
        'epigraph', 'bibliography', 'glossary', 'list-table'
    ]

    def __init__(self, dry_run: bool = False, verbose: bool = False) -> None:
        self.dry_run: bool = dry_run
        self.verbose: bool = verbose
        self.stats: DefaultDict[str, int] = defaultdict(int)

    def convert_file(self, file_path: Union[str, Path]) -> bool:
        """Convert a single file."""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                original_content = f.read()
                original_lines = original_content.splitlines(keepends=True)

            converted_lines = self._convert_lines(original_lines, file_path)
            converted_content = ''.join(converted_lines)

            if converted_content == original_content:
                if self.verbose:
                    print(f"✓ {file_path}: No changes needed")
                return False

            changes = self._count_changes(original_content, converted_content)

            if self.dry_run:
                print(f"🔍 {file_path}: Would convert {changes} directives")
                if self.verbose:
                    self._show_diff(original_lines, converted_lines)
            else:
                with open(file_path, 'w', encoding='utf-8') as f:
                    f.write(converted_content)
                print(f"✅ {file_path}: Converted {changes} directives")

            self.stats['files_modified'] += 1
            self.stats['directives_converted'] += changes
            return True

        except Exception as e:
            print(f"❌ Error processing {file_path}: {e}")
            return False

    def _convert_lines(self, lines: List[str], file_path: Union[str, Path]) -> List[str]:
        """Convert directives in lines."""
        converted = []
        i = 0

        while i < len(lines):
            line = lines[i]

            # Check for colon directive opening
            match = re.match(r'^(\s*):::\{(' + '|'.join(self.BACKTICK_FENCE_DIRECTIVES) + r')\}(.*)$', line)

            if match:
                indent = match.group(1)
                directive_name = match.group(2)
                rest = match.group(3).strip()

                # Convert opening fence - keep rest on same line if present
                if rest:
                    converted.append(f"{indent}```{{{directive_name}}} {rest}\n")
                else:
                    converted.append(f"{indent}```{{{directive_name}}}\n")

                i += 1

                # Process content until closing fence
                while i < len(lines):
                    content_line = lines[i]

                    # Check for closing colon fence
                    closing_match = re.match(r'^(\s*):::\s*$', content_line)
                    if closing_match:
                        # Convert closing fence to backticks (use same or parent indentation)
                        converted.append(f"{indent}```\n")
                        i += 1
                        break
                    else:
                        # Keep content as-is
                        converted.append(content_line)
                        i += 1
            else:
                # Not a directive, keep as-is
                converted.append(line)
                i += 1

        return converted

    def _count_changes(self, original: str, converted: str) -> int:
        """Count number of directives converted."""
        # Count colons in original
        count = 0
        for directive in self.BACKTICK_FENCE_DIRECTIVES:
            count += original.count(f':::{{{directive}}}')
        return count

    def _show_diff(self, original_lines: List[str], converted_lines: List[str]) -> None:
        """Show line-by-line diff."""
        for i, (orig, conv) in enumerate(zip(original_lines, converted_lines), 1):
            if orig != conv:
                print(f"  Line {i}:")
                print(f"    - {orig.rstrip()}")
                print(f"    + {conv.rstrip()}")

def process_directory(content_dir: str, converter: FenceConverter) -> None:
    """Process all markdown files in directory."""
    md_files = []
    for root, dirs, files in os.walk(content_dir):
        for file in files:
            if file.endswith('.md'):
                md_files.append(os.path.join(root, file))

    if not md_files:
        print(f"No markdown files found in {content_dir}")
        return

    print(f"Found {len(md_files)} markdown files to process\n")

    for md_file in sorted(md_files):
        converter.convert_file(md_file)

    print(f"\n=== SUMMARY ===")
    if converter.dry_run:
        print(f"Dry run - no files were modified")
    print(f"Files that need/needed conversion: {converter.stats['files_modified']}")
    print(f"Total directives converted: {converter.stats['directives_converted']}")

def main() -> int:
    parser = argparse.ArgumentParser(
        description='Convert MyST directive fences from backticks to colons'
    )
    parser.add_argument('--dry-run', action='store_true',
                       help='Show changes without applying them')
    parser.add_argument('--content-dir', default='content',
                       help='Content directory to process (default: content)')
    parser.add_argument('--file',
                       help='Convert a single file instead of directory')
    parser.add_argument('--verbose', action='store_true',
                       help='Show detailed conversion info')

    args = parser.parse_args()

    converter = FenceConverter(dry_run=args.dry_run, verbose=args.verbose)

    if args.file:
        if not os.path.exists(args.file):
            print(f"❌ Error: File '{args.file}' not found!")
            return 1
        converter.convert_file(args.file)
    else:
        if not os.path.exists(args.content_dir):
            print(f"❌ Error: Directory '{args.content_dir}' not found!")
            return 1
        process_directory(args.content_dir, converter)

    print("\n✅ Conversion complete!")
    if args.dry_run:
        print("Run without --dry-run to apply changes")
    else:
        print("Run 'myst build' to verify changes")
        print("Run 'python3 scripts/lint_myst_markdown.py' to check remaining issues")

    return 0

if __name__ == "__main__":
    sys.exit(main())

File: scripts/convert_fences.js
#!/usr/bin/env node
/**
 * Bulk Fence Converter for MyST Markdown
 *
 * Converts MyST directives from colon fences (:::) to backtick fences (```)
 * to comply with doc/MYST_CONVENTIONS.md (updated to prefer backticks for IDE support)
 *
 * Status: UTILITY - Occasional use
 *     Run when bulk fence conversion is needed (e.g., after importing content
 *     from other sources that use colon fences).
 *
 * Usage:
 *     node convert_fences.js [--dry-run] [--content-dir DIR] [--file FILE]
 *
 * Options:
 *     --dry-run       Show changes without applying them
 *     --content-dir   Convert all files in directory (default: content)
 *     --file          Convert a single file
 *     --verbose       Show detailed conversion info
 */

const fs = require('fs');
const path = require('path');

// Directives that should use backtick fences
const BACKTICK_FENCE_DIRECTIVES = [
  'note', 'warning', 'important', 'tip', 'caution', 'attention',
  'danger', 'error', 'hint', 'seealso', 'admonition',
  'dropdown', 'card', 'tab-set', 'tab-item',
  'figure', 'subfigure', 'video', 'image',
  'proof', 'theorem', 'lemma', 'definition', 'example', 'exercise',
  'grid', 'column', 'margin', 'sidebar',
  'epigraph', 'bibliography', 'glossary', 'list-table'
];

class FenceConverter {
  /**
   * Converts colon directive fences to backtick fences.
   *
   * @param {boolean} dryRun - Whether to perform a dry run
   * @param {boolean} verbose - Whether to show verbose output
   */
  constructor(dryRun = false, verbose = false) {
    this.dryRun = dryRun;
    this.verbose = verbose;
    this.stats = {
      files_modified: 0,
      directives_converted: 0
    };
  }

  /**
   * Convert a single file.
   *
   * @param {string} filePath - Path to file to convert
   * @returns {boolean} True if file was modified
   */
  convertFile(filePath) {
    try {
      const originalContent = fs.readFileSync(filePath, 'utf8');
      const originalLines = originalContent.split('\n').map(line => line + '\n');

      const convertedLines = this._convertLines(originalLines, filePath);
      const convertedContent = convertedLines.join('');

      if (convertedContent === originalContent) {
        if (this.verbose) {
          console.log(`✓ ${filePath}: No changes needed`);
        }
        return false;
      }

      const changes = this._countChanges(originalContent, convertedContent);

      if (this.dryRun) {
        console.log(`🔍 ${filePath}: Would convert ${changes} directives`);
        if (this.verbose) {
          this._showDiff(originalLines, convertedLines);
        }
      } else {
        fs.writeFileSync(filePath, convertedContent, 'utf8');
        console.log(`✅ ${filePath}: Converted ${changes} directives`);
      }

      this.stats.files_modified++;
      this.stats.directives_converted += changes;
      return true;
    } catch (err) {
      console.log(`❌ Error processing ${filePath}: ${err.message}`);
      return false;
    }
  }

  /**
   * Convert directives in lines.
   *
   * @param {Array<string>} lines - Lines to convert
   * @param {string} filePath - File path for error reporting
   * @returns {Array<string>} Converted lines
   */
  _convertLines(lines, filePath) {
    const converted = [];
    let i = 0;

    const directivePattern = new RegExp(`^(\\s*):::\\{(${BACKTICK_FENCE_DIRECTIVES.join('|')})\\}(.*)$`);

    while (i < lines.length) {
      const line = lines[i];
      const match = line.match(directivePattern);

      if (match) {
        const indent = match[1];
        const directiveName = match[2];
        const rest = match[3].trim();

        // Convert opening fence - keep rest on same line if present
        if (rest) {
          converted.push(`${indent}\`\`\`{${directiveName}} ${rest}\n`);
        } else {
          converted.push(`${indent}\`\`\`{${directiveName}}\n`);
        }

        i++;

        // Process content until closing fence
        while (i < lines.length) {
          const contentLine = lines[i];

          // Check for closing colon fence
          const closingMatch = contentLine.match(/^(\s*):::\s*$/);
          if (closingMatch) {
            // Convert closing fence to backticks (use same or parent indentation)
            converted.push(`${indent}\`\`\`\n`);
            i++;
            break;
          } else {
            // Keep content as-is
            converted.push(contentLine);
            i++;
          }
        }
      } else {
        // Not a directive, keep as-is
        converted.push(line);
        i++;
      }
    }

    return converted;
  }

  /**
   * Count number of directives converted.
   *
   * @param {string} original - Original content
   * @param {string} converted - Converted content
   * @returns {number} Number of directives converted
   */
  _countChanges(original, converted) {
    let count = 0;
    for (const directive of BACKTICK_FENCE_DIRECTIVES) {
      const pattern = `:::{${directive}}`;
      const matches = original.match(new RegExp(pattern, 'g'));
      if (matches) {
        count += matches.length;
      }
    }
    return count;
  }

  /**
   * Show line-by-line diff.
   *
   * @param {Array<string>} originalLines - Original lines
   * @param {Array<string>} convertedLines - Converted lines
   */
  _showDiff(originalLines, convertedLines) {
    const maxLen = Math.max(originalLines.length, convertedLines.length);
    for (let i = 0; i < maxLen; i++) {
      const orig = originalLines[i] || '';
      const conv = convertedLines[i] || '';
      if (orig !== conv) {
        console.log(`  Line ${i + 1}:`);
        console.log(`    - ${orig.trimEnd()}`);
        console.log(`    + ${conv.trimEnd()}`);
      }
    }
  }
}

/**
 * Process all markdown files in directory.
 *
 * @param {string} contentDir - Content directory path
 * @param {FenceConverter} converter - Converter instance
 */
function processDirectory(contentDir, converter) {
  const mdFiles = [];

  function walkDir(dir) {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walkDir(fullPath);
      } else if (entry.name.endsWith('.md')) {
        mdFiles.push(fullPath);
      }
    }
  }

  walkDir(contentDir);

  if (mdFiles.length === 0) {
    console.log(`No markdown files found in ${contentDir}`);
    return;
  }

  console.log(`Found ${mdFiles.length} markdown files to process\n`);

  for (const mdFile of mdFiles.sort()) {
    converter.convertFile(mdFile);
  }

  console.log(`\n=== SUMMARY ===`);
  if (converter.dryRun) {
    console.log(`Dry run - no files were modified`);
  }
  console.log(`Files that need/needed conversion: ${converter.stats.files_modified}`);
  console.log(`Total directives converted: ${converter.stats.directives_converted}`);
}

function main() {
  const args = process.argv.slice(2);
  const options = {
    dryRun: args.includes('--dry-run'),
    contentDir: 'content',
    file: null,
    verbose: args.includes('--verbose')
  };

  // Parse content-dir
  const contentDirIdx = args.indexOf('--content-dir');
  if (contentDirIdx !== -1 && args[contentDirIdx + 1]) {
    options.contentDir = args[contentDirIdx + 1];
  }

  // Parse file
  const fileIdx = args.indexOf('--file');
  if (fileIdx !== -1 && args[fileIdx + 1]) {
    options.file = args[fileIdx + 1];
  }

  const converter = new FenceConverter(options.dryRun, options.verbose);

  if (options.file) {
    if (!fs.existsSync(options.file)) {
      console.log(`❌ Error: File '${options.file}' not found!`);
      return 1;
    }
    converter.convertFile(options.file);
  } else {
    if (!fs.existsSync(options.contentDir)) {
      console.log(`❌ Error: Directory '${options.contentDir}' not found!`);
      return 1;
    }
    processDirectory(options.contentDir, converter);
  }

  console.log('\n✅ Conversion complete!');
  if (options.dryRun) {
    console.log('Run without --dry-run to apply changes');
  } else {
    console.log('Run \'myst build\' to verify changes');
    console.log('Run \'node scripts/lint_myst_markdown.js\' to check remaining issues');
  }

  return 0;
}

if (require.main === module) {
  process.exit(main());
}

module.exports = { FenceConverter, processDirectory };
